tabular/src/autogluon/tabular/learner/default_learner.py

```diff
@@ -241,107 +236,68 @@ def general_data_processing(self, X: DataFrame, X_val: DataFrame, X_unlabeled: D
-                y_val = self.label_cleaner.transform(y_val)
-                X_val = self.set_predefined_weights(X_val, y_val)
-                X_val, w_val = extract_column(X_val, self.sample_weight)
-        else:
-            y_val = None
-            w_val = None
+        X_val, y_val, w_val, holdout_frac = self._apply_cleaner_transform(X_val, y_uncleaned, holdout_frac, holdout_frac_og, "X_val")
+        X_test, y_test, w_test, _ = self._apply_cleaner_transform(X_test, y_uncleaned, holdout_frac, holdout_frac_og, "X_test")
 
         self._original_features = list(X.columns)
         # TODO: Move this up to top of data before removing data, this way our feature generator is better
         logger.log(20, f"Using Feature Generators to preprocess the data ...")
-        if X_val is not None:
-            # Do this if working with SKLearn models, otherwise categorical features may perform very badly on the test set
+
+        if X_val is not None or X_test is not None:
             logger.log(
                 15,
-                "Performing general data preprocessing with merged train & validation data, so validation performance may not accurately reflect performance on new test data",
+                "Performing general data preprocessing with merged train & validation data, so validation/test performance may not accurately reflect performance on new test data",
             )
-            X_super = pd.concat([X, X_val, X_unlabeled], ignore_index=True)
-            if self.feature_generator.is_fit():
-                logger.log(
-                    20,
-                    f"{self.feature_generator.__class__.__name__} is already fit, so the training data will be processed via .transform() instead of .fit_transform().",
-                )
-                X_super = self.feature_generator.transform(X_super)
-                self.feature_generator.print_feature_metadata_info()
-            else:
-                if X_unlabeled is None:
-                    y_super = pd.concat([y, y_val], ignore_index=True)
-                else:
-                    y_unlabeled = pd.Series(np.nan, index=X_unlabeled.index)
-                    y_super = pd.concat([y, y_val, y_unlabeled], ignore_index=True)
-                X_super = self.fit_transform_features(X_super, y_super, problem_type=self.label_cleaner.problem_type_transform, eval_metric=self.eval_metric)
-            X = X_super.head(len(X)).set_index(X.index)
-
-            X_val = X_super.head(len(X) + len(X_val)).tail(len(X_val)).set_index(X_val.index)
-
-            if X_unlabeled is not None:
-                X_unlabeled = X_super.tail(len(X_unlabeled)).set_index(X_unlabeled.index)
-            del X_super
+
+        datasets = [X, X_val, X_test, X_unlabeled]
+        X_super = pd.concat(datasets, ignore_index=True)
+
+        y_unlabeled = pd.Series(np.nan, index=X_unlabeled.index) if X_unlabeled is not None else None
+        y_list = [y, y_val, y_test, y_unlabeled]
```

Comment:
MAJOR ISSUE: I noticed this now and it might be why the training differs between passing `test` and not passing `test`.

`test` can impact the preprocessing logic.

We do not want to use `X_test` in the `fit_transform_features` logic. Instead, we want to transform test separately. via a `transform` call after fit.

Reply:
Instead we should do:

```
X_test = self.feature_generator.transform(X_test)
```

Reply:
We should keep both sets of logic and have a flag to enable one vs the other. The default should be

```
X_test = self.feature_generator.transform(X_test)
```